From: Simon McVittie <smcv@debian.org>
Date: Tue, 6 Sep 2016 11:20:39 +0100
Subject: document portal: cope with multiple events that would cause failure

If the name is lost *and* the session bus is closed, we would
crash with an assertion failure when the GError is overwritten.

Signed-off-by: Simon McVittie <smcv@debian.org>
---
 document-portal/xdp-main.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/document-portal/xdp-main.c b/document-portal/xdp-main.c
index 990c6f8..20b8641 100644
--- a/document-portal/xdp-main.c
+++ b/document-portal/xdp-main.c
@@ -1009,8 +1009,13 @@ on_name_lost (GDBusConnection *connection,
               gpointer         user_data)
 {
   g_debug ("%s lost", name);
-  final_exit_status = 20;
-  g_set_error (&exit_error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED, "D-Bus name \"%s\" lost", name);
+
+  if (final_exit_status == 0)
+    final_exit_status = 20;
+
+  if (exit_error == NULL)
+    g_set_error (&exit_error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED, "D-Bus name \"%s\" lost", name);
+
   g_main_loop_quit (loop);
 }
 
@@ -1028,7 +1033,9 @@ session_bus_closed (GDBusConnection *connection,
                     gboolean         remote_peer_vanished,
                     GError          *bus_error)
 {
-  g_set_error (&exit_error, G_IO_ERROR, G_IO_ERROR_BROKEN_PIPE, "Disconnected from session bus");
+  if (exit_error == NULL)
+    g_set_error (&exit_error, G_IO_ERROR, G_IO_ERROR_BROKEN_PIPE, "Disconnected from session bus");
+
   g_main_loop_quit (loop);
 }
 
